### ISSUE TITLE
Discussion regarding the BDF solver

### ISSUE THREAD

**Issue by jiweiqi** [comocheng/jax-reactor#5]
The BDF solver is really cool! I have a quick question regarding the AD of BDF solver.

Is it possible to use Auto-Differentiation across the BDF solver, so that we can do efficient adjoint sensitivity analysis?

For adjoint sensitivity analysis, we have the species concentrations `Y = F(t)`, and we define a loss function `g(Y)` and compute the gradient of `g(Y)` for optimizing the mechanism.

**Comment by skrsna** [comment]
Hi @jiweiqi ,

Thanks for asking. I'm currently to trying to write a `custom_vjp` using an adjoint method similar to this one https://github.com/google/jax/blob/master/jax/experimental/ode.py#L265. Turns out this is not that trivial with BDF solver cause there are lots of constraints in JAX. I'm hoping to get some feedback from JAX's developers on the PR (https://github.com/google/jax/pull/3781) and finish writing it.

**Comment by jiweiqi** [comment]
I just read the Jax PR. Regarding the slowness, you might take a look at the clipping and normalization of the mass fraction. Clipping and normalization of the mass fraction could affect the properties of the Jacobian matrix. This has been discussed somewhere else in the Cantera google group. For ReacTorch, we are working on fixing this [issue](https://github.com/DENG-MIT/reactorch/issues/40).

**Comment by skrsna** [comment]
Thanks for the suggestion. I think this is what you are talking about https://github.com/comocheng/jax-reactor/blob/master/jax_reactor/solution.py#L79. I might be wrong but let me know!!

**Comment by jiweiqi** [comment]
Yes. Besides that, another clip is in the wdot function https://github.com/comocheng/jax-reactor/blob/e13d359c6171bd1c41ede244570380184cb54870/jax_reactor/solution.py#L58.

**Comment by jiweiqi** [comment]
I am not sure if this is the case. An easy way to test it is to print out the number of function evaluations and jacobian evaluations. If you see significantly more computation than Cantera, then that is probably related to this issue.
